File: frontend/src/app/core/models/game.models.ts
export interface Position { x: number; y: number; }

export type TileType = 'plains' | 'forest' | 'hills' | 'mountains' | 'ocean' | 'desert' | 'tundra' | 'river';
export type ResourceType = 'iron' | 'cattle' | 'luxury_silk' | 'wheat' | 'coal' | 'oil';
export type UnitType = 'warrior' | 'archer' | 'knight' | 'settler' | 'tank' | 'eagle_warrior' | 'legion';
export type BuildingType = 'granary' | 'library' | 'barracks' | 'market' | 'temple' | 'forge' | 'palace' | 'workshop' | 'aqueduct' | 'colosseum' | 'university' | 'factory' | 'power_plant' | 'space_shuttle';
export type DiplomacyStatus = 'peace' | 'war' | 'alliance';
export type GameResult = 'ongoing' | 'player_win_conquest' | 'player_win_science' | 'player_win_culture' | 'player_win_domination' | 'ai_win_conquest' | 'ai_win_science' | 'ai_win_culture' | 'ai_win_domination';

export interface Resources {
  food: number;
  production: number;
  science: number;
  gold: number;
  culture: number;
  happiness: number;
}

export interface Tile {
  x: number;
  y: number;
  tile_type: TileType;
  resource?: ResourceType;
  resource_improved: boolean;
  has_ruins: boolean;
  has_barbarian_camp: boolean;
  road: boolean;
  owner?: string;
}

export interface Building {
  building_type: BuildingType;
  turns_built: number;
}

export interface ProductionQueue {
  item_type: string;
  is_unit: boolean;
  turns_remaining: number;
}

export interface City {
  id: string;
  name: string;
  owner: string;
  position: Position;
  population: number;
  food_stored: number;
  buildings: Building[];
  production_queue?: ProductionQueue;
  yields: Resources;
  happiness: number;
  culture: number;
}

export interface Unit {
  id: string;
  unit_type: UnitType;
  owner: string;
  position: Position;
  movement_points: number;
  movement_points_left: number;
  strength: number;
  health: number;
  promotions: string[];
  has_attacked: boolean;
}

export interface MapData {
  width: number;
  height: number;
  tiles: Tile[];
  fog_of_war: boolean[][];
}

export interface PlayerState {
  civ_id: string;
  resources: Resources;
  cities: City[];
  units: Unit[];
  researched_techs: string[];
  current_research?: string;
  research_progress: number;
  diplomacy_status: DiplomacyStatus;
  score: number;
  spaceship_built: boolean;
}

export interface GameState {
  id: string;
  name: string;
  turn: number;
  current_phase: string;
  current_player: 'player' | 'ai';
  player: PlayerState;
  ai: Partial<PlayerState> & { cities: any[]; units: any[] };
  map: MapData;
  result: GameResult;
  max_turns: number;
  cheats_used: string[];
}

export interface Scenario {
  id: string;
  name: string;
  description: string;
  difficulty: string;
  map_width: number;
  map_height: number;
}

export interface SavedGameSummary {
  id: string;
  name: string;
  scenario_id: string;
  created_at: string;
  last_saved: string;
  is_autosave: boolean;
  turn: number;
  result: string;
}

export interface ActionRequest {
  action_type: string;
  details: Record<string, unknown>;
}

File: backend/models/game_models.py
"""
Pydantic models for the game domain.
All MongoDB documents use str(_id) serialized as 'id'.
"""
from __future__ import annotations
from typing import Any, Optional
from pydantic import BaseModel, Field
from enum import Enum


# ─── Enums ────────────────────────────────────────────────────────────────────

class TileType(str, Enum):
    PLAINS = "plains"
    FOREST = "forest"
    HILLS = "hills"
    MOUNTAINS = "mountains"
    OCEAN = "ocean"
    DESERT = "desert"
    TUNDRA = "tundra"
    RIVER = "river"


class ResourceType(str, Enum):
    IRON = "iron"
    CATTLE = "cattle"
    LUXURY_SILK = "luxury_silk"
    WHEAT = "wheat"
    COAL = "coal"
    OIL = "oil"


class UnitType(str, Enum):
    WARRIOR = "warrior"
    ARCHER = "archer"
    KNIGHT = "knight"
    SETTLER = "settler"
    TANK = "tank"           # cheat / advanced tech unlock
    UNIQUE_PLAYER = "eagle_warrior"
    UNIQUE_AI = "legion"


class BuildingType(str, Enum):
    GRANARY = "granary"
    LIBRARY = "library"
    BARRACKS = "barracks"
    MARKET = "market"
    TEMPLE = "temple"
    FORGE = "forge"
    PALACE = "palace"       # unique – capital auto-built


class TechEra(str, Enum):
    ANCIENT = "ancient"
    CLASSICAL = "classical"
    MODERN = "modern"


class DiplomacyStatus(str, Enum):
    PEACE = "peace"
    WAR = "war"
    ALLIANCE = "alliance"


class TurnPhase(str, Enum):
    PRODUCTION = "production"
    RESEARCH = "research"
    MOVEMENT = "movement"
    DIPLOMACY = "diplomacy"
    BUILDING = "building"
    END = "end"


class GameResult(str, Enum):
    ONGOING = "ongoing"
    PLAYER_WIN_CONQUEST = "player_win_conquest"
    PLAYER_WIN_SCIENCE = "player_win_science"
    PLAYER_WIN_CULTURE = "player_win_culture"
    PLAYER_WIN_DOMINATION = "player_win_domination"
    AI_WIN_CONQUEST = "ai_win_conquest"
    AI_WIN_SCIENCE = "ai_win_science"
    AI_WIN_CULTURE = "ai_win_culture"
    AI_WIN_DOMINATION = "ai_win_domination"


# ─── Sub-models ───────────────────────────────────────────────────────────────

class Position(BaseModel):
    x: int
    y: int


class Resources(BaseModel):
    food: int = 0
    production: int = 0
    science: int = 0
    gold: int = 0
    culture: int = 0
    happiness: int = 0


class Tile(BaseModel):
    x: int
    y: int
    tile_type: TileType = TileType.PLAINS
    resource: Optional[ResourceType] = None
    resource_improved: bool = False
    has_ruins: bool = False
    has_barbarian_camp: bool = False
    road: bool = False
    # Ownership: "player", "ai", or None
    owner: Optional[str] = None


class Building(BaseModel):
    building_type: BuildingType
    turns_built: int = 0  # track when it was completed


class ProductionQueue(BaseModel):
    item_type: str  # BuildingType or UnitType value
    is_unit: bool = False
    turns_remaining: int = 1


class City(BaseModel):
    id: str
    name: str
    owner: str  # "player" or "ai"
    position: Position
    population: int = 1
    food_stored: int = 0
    buildings: list[Building] = []
    production_queue: Optional[ProductionQueue] = None
    # Per-turn yields (computed by resource calculator)
    yields: Resources = Field(default_factory=Resources)
    happiness: int = 5
    culture: int = 0


class Unit(BaseModel):
    id: str
    unit_type: UnitType
    owner: str  # "player" or "ai"
    position: Position
    movement_points: int = 2
    movement_points_left: int = 2
    strength: int = 10
    health: int = 100
    promotions: list[str] = []
    has_attacked: bool = False


class Technology(BaseModel):
    tech_id: str
    name: str
    era: TechEra
    cost: int  # science points
    required_techs: list[str] = []
    unlocks_buildings: list[BuildingType] = []
    unlocks_units: list[UnitType] = []
    culture_bonus: int = 0
    production_bonus: int = 0


class CivilizationDef(BaseModel):
    civ_id: str
    name: str
    unique_unit: UnitType
    science_bonus: float = 1.0
    production_bonus: float = 1.0
    culture_bonus: float = 1.0
    gold_bonus: float = 1.0
    starting_techs: list[str] = []


class PlayerState(BaseModel):
    civ_id: str
    resources: Resources = Field(default_factory=Resources)
    cities: list[City] = []
    units: list[Unit] = []
    researched_techs: list[str] = []
    current_research: Optional[str] = None
    research_progress: int = 0
    diplomacy_status: DiplomacyStatus = DiplomacyStatus.PEACE
    score: int = 0
    spaceship_built: bool = False


class MapData(BaseModel):
    width: int
    height: int
    tiles: list[Tile] = []
    # fog_of_war[y][x] = True means explored by player
    fog_of_war: list[list[bool]] = []


class GameState(BaseModel):
    turn: int = 1
    current_phase: TurnPhase = TurnPhase.MOVEMENT
    current_player: str = "player"  # "player" or "ai"
    player: PlayerState
    ai: PlayerState
    map: MapData
    result: GameResult = GameResult.ONGOING
    max_turns: int = 200
